### PR TITLE
Use hyper-openssl instead of hyper-tls

### LINKED ISSUE #66: Outdated OpenSSL and hyper-tls
As per @CodeSandwich in #51 

> I'd suggest upgrading yet another dependency, which is burried deep in other outdated crates: openssl. Current version (0.9.24) does not compile on systems with OpenSSL 1.1.1.

Sadly we can't update OpenSSL to a later version.

In native-tls 0.2, which supports Rust OpenSSL 0.10, they've dropped support for `TlsConnectorBuilderExt::from_openssl`, which we use in https://github.com/Metaswitch/swagger-rs/blob/f86922a3b9a6967831faa891922f0bedf38d0fec/src/connector.rs#L88-L99

Until we add sufficient function to native TLS to do what we are doing by using OpenSSL manually, we can't move.

I've raised https://github.com/sfackler/rust-native-tls/issues/125 to look at getting a fix to unblock this - until then, this issue is blocked.

### PR DESCRIPTION
**Pull request by richardwhiuk**
Fixes #66 and #43

We move to hyper-openssl on Linux, so we can update OpenSSL, and maintain support for pinning certificates and mutual TLS.

We stick with hyper-tls and native-tls on non-Linux platforms, but don't support that feature (as it's not supported in native-tls in the same fashion).

### CODE REVIEW

**Comment by bnewbold** [comment]
Hi, thanks for maintaining `swagger-rs`!

Is there anything I can do to help land this particular PR? It seems like it will resolve the OpenSSL 1.0/1.1 dependency issue on Linux (at least for my use case).

**Comment by richardwhiuk** [comment]
@bossmc Can you review this?

File: src/connector.rs [hunk 1]
```diff
@@ -1,114 +1,144 @@
...
+        self
+    }
+
+    /// Pin the CA certificate for the server's certificate.
+    ///
+    /// # Arguments
+    ///
+    /// * `ca_certificate` - Path to CA certificate used to authenticate the server
+    #[cfg(not(any(target_os = "macos", target_os = "windows", target_os = "ios")))]
+    pub fn pin_server<CA>(mut self, ca_certificate: CA) -> Self
```
Comment (bossmc) [inline src/connector.rs]: Ditto (to the below) - `pin_server_cert`?

File: src/connector.rs [hunk 2]
```diff
@@ -1,114 +1,144 @@
...
-    _client_certificate: C,
-) -> Box<dyn Fn() -> hyper_tls::HttpsConnector<hyper::client::HttpConnector> + Send + Sync>
-where
-    CA: AsRef<Path>,
-    K: AsRef<Path>,
-    C: AsRef<Path>,
-{
-    unimplemented!("See issue #43 (https://github.com/Metaswitch/swagger-rs/issues/43)")
+    #[cfg(any(target_os = "macos", target_os = "windows", target_os = "ios"))]
+    /// Build the HTTPS connector. Will if the SSL connector can't be created.
```
Comment (bossmc) [inline src/connector.rs]: Will what?

File: CHANGELOG.md [hunk 3]
```diff
@@ -8,8 +8,16 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ### Added
 
 ### Changed
+- Use hyper-openssl on Linux, instead of hyper-tls
+- Use a builder pattern to created client connectors
+- Allow HTTPS connectors to be built which don't pin the server CA certificate
+- Allow HTTPS to work on Mac/Windows/iOS
```
[the remark this answers is not in the document]
Comment (richardwhiuk) [reply CHANGELOG.md]: Nope - https://github.com/Metaswitch/swagger-rs/pull/79/files#diff-0bb6ab7e9f36ebdf213bd0597dd33954R132-R133 adds HTTPS support Mac OS X, iOS and Windows using `hyper-tls`.

File: CHANGELOG.md [hunk 4]
```diff
@@ -8,8 +8,16 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- Enforce that HTTPS is used if we are using a HTTPS connector.
+- Return Results, rather than unwrapping errors on client creation
+- openssl 0.10
 
 ### Removed
+- Remove HTTPS methods for Mac OS X, iOS and Windows
```
Comment (bossmc) [inline CHANGELOG.md]: Then I question this line in the CHANGELOG....

Comment (richardwhiuk) [reply CHANGELOG.md]: Technically true, I removed them on all patterns, and replaced with a builder pattern - but I'll tidy this up.

**Comment by bossmc** [comment]
:shipit:

**Comment by bossmc** [comment]
bors delegate+

**Pull request merged**